api/v1beta1/machinedeployment_types.go

```diff
@@ -67,10 +71,18 @@ type MachineDeploymentSpec struct {
 	// +kubebuilder:validation:MinLength=1
 	ClusterName string `json:"clusterName"`
 
-	// Number of desired machines. Defaults to 1.
+	// Number of desired machines.
 	// This is a pointer to distinguish between explicit zero and not specified.
+	//
+	// Defaults to:
+	// * if the DefaultReplicasAnnotation annotation is set, use the annotation value
```

Comment:
How does the annotation help to "mprove replica defaulting for autoscaler"?

Reply:
Let's assume you create a new MD which should be controlled by the autoscaler with something that continuously writes the MD (ClusterClass, GitOps, kubectl apply continuously).

In this case the MD does not have the replica field set and before this PR replicas was always defaulted to 1.

If you have autoscaler running with the default annotations it's fine as the webhook automatically picks the default based on min-size.

If you run the autoscaler with different annotations the autoscaler only takes over if the (min size, max size) range includes 1. If you want to have a higher min size the autoscaler will never take control as 1 is outside of the range.

So essentially you can use the default annotation if for whatever reason the min-size, max-size annotations don't work (because autoscaler is running with a different CAPI_GROUP, or because a different autoscaler is used that might have the same limitations as the cluster-autoscaler).

Reply:
ah I see.
In Openshift the autoscaling annotation are not consumer facing. Are gated by operators, the annotations are impl details and so they should be transitioned in backward compatible manner regardless of the autoscaler operating over openshift.io or cluster.x-k8s.io.

As for "if for whatever reason the min-size, max-size annotations don't work..." I think we should wait to hear from real scenarios.
My preference would be to drop the annotation for now until there's a clear/stronger need.

Reply:
@fabriziopandini What do you think?

If I remember correctly, all the use cases we wanted to cover work with min size and max size annotations alone and we only added the default-replicas one to support folks who might run the cluster-autoscaler with non-default annotation names.

Reply:
> As for "if for whatever reason the min-size, max-size annotations don't work..." I think we should wait to hear from real scenarios.

this makes a lot of sense to me. maybe we should wait to see if it an issue that users will face?

Reply:
I'm also ok in dropping the “machinedeployment.cluster.x-k8s.io/default-replicas” for now and eventually iterate back if we have more clear requirements, let's make sure to highlight the boundaries of the current solution (default node group name, default annotations names, only K8s autoscaler)

Reply:
I dropped the default-replicas annotation

Reply:
Where do I highlight the boundaries of the current solution?